readme.md

```diff
@@ -156,24 +156,45 @@ An example would be an `Authorization` header with a signature depending on host
 While this could be calculated before making the request it is just convenient to have it done
 automatically with each request.
 
-For this purpose `Proc`s/`lambda`s can be set on the Grac object following the example below:
+For this purpose an object can be added as middleware which supports the methods `chain` and `call`.
+`chain` is used to chain other middleware down to the actual request. It accepts one parameter which
+should be the `Grac` object or another middleware and it should return itself.
+`call` should accept the parameters as shown in the example below:
 
 ```ruby
-mw = lambda do |opts, request_uri, method, params, body|
-  # your code here
-  # opts        - Hash of the options currently set on the grac object
-  # request_uri - uri returned by grac
-  # method      - http method (lower case)
-  # params      - hash of all params for this request
-  # body        - serialized body
-  return opts, request_uri, method, params, body
+class MW
+  def initialize(options = {})
+    # Use initialize to add any specific configuration
+  end
+
+  def chain(request)
+    @request = request
+    return self
+  end
```

Comment:
`chain` does something we generally try to avoid in Grac: Mutate an object's state.

I guess you want to pass some variables to the middleware. Instead of changing the middleware using the chain method, I'd suggest adopting a pattern similar to Rack, where Rack instantiates the middleware and passes all arguments after the middleware's name in a `use`-command to the middleware's initializer.

That would also allow caching a middleware chain - we create a new Grac object if anything changes anyway. I'm not sure whether it makes more sense to do this lazily or when creating the object. I guess the latter would be cleaner.